File: apps/flutter-prototype/test/recovery_chat_sink_test.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter_test/flutter_test.dart';
import 'package:meshx_flutter_probe/application/recovery_chat_sink.dart';
import 'package:meshx_flutter_probe/core/models.dart';
import 'package:meshx_flutter_probe/core/recovery.dart';
import 'package:meshx_flutter_probe/core/recovery_data.dart';
import 'package:meshx_flutter_probe/core/recovery_snapshot.dart';
import 'package:meshx_flutter_probe/platform/storage.dart';

void main() {
  test(
    'full rebuild removes absent bodies and separates held outbox atomically',
    () async {
      final root = await Directory.systemTemp.createTemp('meshx-sink-');
      addTearDown(() => root.delete(recursive: true));
      final store = FileChatStore(directory: root);
      final f =
          jsonDecode(
                File(
                  '../../contracts/test-vectors/recovery-snapshot-pages.json',
                ).readAsStringSync(),
              )
              as Map;
      (f['message']['details'] as Map).addAll(<String, dynamic>{
        'isBurn': 1,
        'burnDuration': 60,
        'replyToId': 'reply-a',
        'mentionUserIds': '8,9',
      });
      final c = f['context'];
      final context = RecoveryContext(
        origin: c['origin'],
        userId: c['userId'],
        streamEpoch: c['streamEpoch'],
        generation: c['generation'],
      );
      ChatMessage pending(String cid, String body) => ChatMessage(
        messageId: '',
        clientMsgId: 'pending-$cid',
        conversationId: cid,
        fromUserId: 7,
        content: body,
        sequence: 0,
        createdAt: DateTime(2026),
        delivery: Delivery.queued,
      );
      RecoveredChatImage? published;
      final sink = RecoveryChatSink(
        store: store,
        seed: () => RecoveryChatData(
          {
            'group:99': [pending('group:99', 'revoked-secret')],
            'group:21': [pending('group:21', 'draft')],
          },
          [],
          {},
        ),
        isCurrent: (_) => true,
        onPublish: (_, image) => published = image,
        onQuarantine: (_) {},
      );
      await store.save('https://node.example|7', {'content': 'legacy-secret'});
      await sink.begin(context);
      final result = stageRecoverySnapshotPage(
        beginFullRecoverySnapshot(context, 's', '0'),
        context,
        null,
        {
          'snapshotId': 's',
          'boundary': '0',
          'items': [f['directory'], f['message']],
          'nextPageToken': null,
          'snapshotComplete': true,
        },
      );
      await sink.snapshot(context, result.items);
      final ready = result.stage.state.copyWith(
        phase: RecoveryPhase.onlineSafe,
      );
      expect(
        () => sink.publish(context, ready),
        throwsA(isA<RecoveryProtocolError>()),
      );
      await sink.commit(context, result.stage.state);
      sink.publish(context, ready);
      expect(published!.chat.messages['group:21'], hasLength(1));
      expect(published!.outbox['group:21'], hasLength(1));
      final disk = (await FileChatStore(
        directory: root,
      ).loadRecovery('https://node.example|7'))!.snapshot;
      expect(jsonEncode(disk), isNot(contains('revoked-secret')));
      expect(jsonEncode(disk), isNot(contains('legacy-secret')));
      expect((disk['messages'] as Map)['group:21'], hasLength(1));
      final normal = ChatMessage.restore(
        Map<String, dynamic>.from(disk['messages']['group:21'][0]),
      );
      expect(normal.isBurn, isTrue);
      expect(normal.burnDuration, 60);
      expect(normal.replyToId, 'reply-a');
      expect(
        normal
            .withRecoveryDisposition(RecoveryDisposition.needsUserAction)
            .mentionUserIds,
        '8,9',
      );
      expect(
        normal
            .withoutRecoveryBody(isRecalled: true)
            .toJson()
            .containsKey('replyToId'),
        isFalse,
      );
      final outbox = disk['outbox'] as Map;
      expect(outbox['group:99'][0]['recoveryDisposition'], 'dropBodyRevoked');
      expect(outbox['group:21'][0]['content'], 'draft');
      expect(outbox['group:21'][0]['delivery'], 'failed');
      expect(disk['recovery']['cursor'], '0');
      final restarted = RecoveryChatSink(
        store: FileChatStore(directory: root),
        seed: () => RecoveryChatData({}, [], {}),
        isCurrent: (_) => true,
        onPublish: (_, _) {},
        onQuarantine: (_) {},
      );
      await restarted.begin(context);
      await restarted.snapshot(context, result.items);
      await restarted.commit(context, result.stage.state);
      final savedAgain = (await store.loadRecovery(
        'https://node.example|7',
      ))!.snapshot;
      expect(savedAgain['outbox']['group:21'][0]['content'], 'draft');
      expect(
        savedAgain['outbox']['group:99'][0]['recoveryDisposition'],
        'dropBodyRevoked',
      );
      sink.quarantine('DISCONNECTED');
      await expectLater(
        sink.commit(context, result.stage.state),
        throwsA(isA<RecoveryProtocolError>()),
      );
      expect((await store.loadRecovery('https://node.example|7'))!.revision, 2);
    },
  );
}

File: apps/web/tests/recovery-chat-sink.test.mjs
import assert from 'node:assert/strict'
import {readFileSync} from 'node:fs'
import test from 'node:test'
import {createTsLoader,plain,deferred} from './helpers/load-ts.mjs'
const load=createTsLoader({replacements:{'apps/web/src/services/localChatDb.ts':{}}})
const {RecoveryChatSink}=load('apps/web/src/services/recoveryChatSink.ts')
const {RecoveryCoordinator}=load('apps/web/src/services/recoveryCoordinator.ts')
const f=JSON.parse(readFileSync(new URL('../../../contracts/test-vectors/recovery-snapshot-pages.json',import.meta.url)))
function harness(fail=false) {
  let disk, published, readyCalls=0, revision=0, seeded=true
  const queue=cid=>({conversationId:cid,clientMsgId:`pending-${cid}`,requestId:'request',createdAt:'2026-09-14T00:00:00Z',retryCount:0,state:'WAITING_NETWORK',payload:{contentType:'text',content:`draft-${cid}`,isBurn:false}})
  const sink=new RecoveryChatSink(()=>({outbox:seeded?[queue('group:99'),queue('group:21')]:[]}),()=>true,(_,image)=>published=image,()=>{},
    {loadRecoverySnapshot:async()=>disk?{revision,snapshot:disk}:null,commitRecoverySnapshot:async(_,expected,snapshot)=>{assert.equal(expected,revision);if(fail)throw Error('disk full');disk=plain(snapshot);return ++revision}})
  const transport={capabilities:async()=>({capability:'meshx.mutation-recovery',versions:[1],recordVersion:1,maxPageSize:100}),
    open:async()=>({recoveryId:'session',mode:'rebuild',streamEpoch:f.context.streamEpoch,startCursor:'0',snapshotBoundary:'0',floor:'0',latest:'0',snapshotId:'snapshot'}),
    snapshot:async()=>({snapshotId:'snapshot',boundary:'0',items:[f.directory,f.message],nextPageToken:null,snapshotComplete:true}),
    cut:async()=>({through:'1',floor:'0',streamEpoch:f.context.streamEpoch}),
    mutations:async()=>({fromExclusive:'0',through:'1',nextCursor:'1',hasMore:false,floor:'0',latest:'1',streamEpoch:f.context.streamEpoch,
      records:[{recordVersion:1,eventId:'22222222-2222-4222-8222-222222222222',streamEpoch:f.context.streamEpoch,cursor:'1',type:'MESSAGE_RECALLED',conversationId:'group:21',messageId:'message-a',objectVersion:'2',committedAt:'2026-09-14T00:00:00.000Z'}]}),
    ready:async()=>{readyCalls++;assert.ok(disk);return {ready:true,acceptedCursor:'1',latest:'1',streamEpoch:f.context.streamEpoch}},release:async()=>{}}
  return {runner:new RecoveryCoordinator(transport,sink,()=>true),sink,disk:()=>disk,published:()=>published,readyCalls:()=>readyCalls,restart:()=>{seeded=false}}
}
test('coordinator and actual model sink remove terminal bodies before durable READY',async()=>{
  const h=harness(),result=await h.runner.rebuild(f.context,'key')
  assert.equal(result.phase,'ONLINE_SAFE');assert.equal(h.readyCalls(),1)
  const disk=h.disk()
  assert.equal(disk.recovery.cursor,'1');assert.equal(disk.messages.length,0)
  assert.equal(disk.tombstones[0].state,'RECALLED');assert.equal(disk.tombstones[0].messageSequence,'1')
  assert.equal(JSON.stringify(disk).includes('fixture body'),false)
  assert.equal(JSON.stringify(disk).includes('draft-group:99'),false)
  assert.equal(disk.outbox[0].recoveryDisposition,'DROP_BODY_REVOKED')
  assert.equal(disk.outbox[1].recoveryDisposition,'NEEDS_USER_ACTION');assert.equal(disk.outbox[1].payload.content,'draft-group:21')
  assert.deepEqual(plain(h.published().messages),[])
  h.restart()
  assert.equal((await h.runner.rebuild(f.context,'restart-key')).phase,'ONLINE_SAFE')
  assert.equal(h.disk().outbox.length,2)
  assert.equal(h.disk().outbox[1].payload.content,'draft-group:21')
  h.sink.quarantine('DISCONNECTED')
  assert.throws(()=>h.sink.publish(f.context,result),/STALE_RECOVERY/)
})
test('failed actual sink commit never sends READY or publishes',async()=>{
  const h=harness(true),result=await h.runner.rebuild(f.context,'key')
  assert.equal(result.phase,'STORAGE_BLOCKED');assert.equal(h.readyCalls(),0)
  assert.equal(h.disk(),undefined);assert.equal(h.published(),undefined)
})

test('same-owner rebuild invalidates late commit completion from the previous run',async()=>{
  const gate=deferred(),entered=deferred()
  const {initialRecoveryState}=load('packages/domain-ts/src/recovery.ts')
  const state={...initialRecoveryState(f.context),phase:'CATCHING_UP'}
  const sink=new RecoveryChatSink(()=>({outbox:[]}),()=>true,()=>{},()=>{},
    {loadRecoverySnapshot:async()=>null,commitRecoverySnapshot:async()=>{entered.resolve();await gate.promise;return 1}})
  await sink.begin(f.context)
  const old=sink.commit(f.context,state)
  const rejection=assert.rejects(old,/STALE_RECOVERY/)
  await entered.promise
  sink.quarantine('DISCONNECTED');await sink.begin(f.context)
  gate.resolve();await rejection
  assert.throws(()=>sink.publish(f.context,{...state,phase:'ONLINE_SAFE'}),/PROTOCOL_ERROR/)
})
